### PR TITLE
Kill list implementation with a patch route

### PR DESCRIPTION
**Pull request by Rackover**
On iw4x - and maybe other clients - people use the master server to advertise their discord with locked servers or fake player count, we would like to have a way to handle these issues easily by banning target IPs from the master server.

This should be extremely simple to do and require no redeploy/pr/snake intervention, cause if it takes more effort to update the kill list than it takes malicious hosters to change their IP, it won't be useful.

So I implemented a very simple kill list TXT the server writes to and reads from, and that can be very easily updated with a patch command sent with a secret key (stored in ENV in serverside). Ultimately we could have the XLabs discord bot send that message on demand (so the bot would have the secret key), so that we could in the admin channel update the kill list with just a single discord message. Very neat!

The kill list also supports adding a reason to any kill line so that we can remember why a certain ip is removed or not.

**Note: This builds and runs, but I have yet to test it in different edge cases to make sure it is 100% OK**

### CODE REVIEW

**Comment by Emosewaj** [comment]
I maintain the discord bot, I'd be happy to include a slash-command for this once it goes through. If this gets merged, hit me up on discord so we can talk specifics.

File: src/services/kill_list.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		auto entry = entries.at(str_address);
+
+		reason = entry.reason;
+		return true;
+	}
+
+	return false;
+}
+
+void kill_list::add_to_kill_list(const kill_list::kill_list_entry& add)
```
Comment (momo5502) [inline src/services/kill_list.cpp:24]: const ref could be a moved object instead

File: src/services/patch_kill_list_command.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+void patch_kill_list_command::handle_command([[maybe_unused]] const network::address& target, const std::string_view& data)
+{
+	std::string key;
+	size_t size;
+
+#if _WIN32
+	char key_buff[128];
+	auto result = getenv_s(&size, key_buff, 128, key_env_name.data());
```
Comment (momo5502) [inline src/services/patch_kill_list_command.cpp:21]: why env and not a text file next to it, just like patreon key is read?

File: src/services/kill_list.hpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	void add_to_kill_list(const kill_list::kill_list_entry& add);
+	void remove_from_kill_list(const network::address& remove);
+	void remove_from_kill_list(const std::string& remove);
+
+private:
+	std::string secret_;
+	std::unordered_map<std::string, kill_list_entry> entries;
+	void reload_from_disk();
+	void write_to_disk();
+	std::recursive_mutex kill_list_mutex;
```
Comment (momo5502) [inline src/services/kill_list.hpp:37]: why recursive?

File: src/services/kill_list.hpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	kill_list(server& server);
+
+	bool contains(const network::address& address, std::string& reason);
+	void add_to_kill_list(const kill_list::kill_list_entry& add);
+	void remove_from_kill_list(const network::address& remove);
+	void remove_from_kill_list(const std::string& remove);
+
+private:
+	std::string secret_;
+	std::unordered_map<std::string, kill_list_entry> entries;
```
Comment (momo5502) [inline src/services/kill_list.hpp:34]: there is a concurrency container class that wraps a container together with a mutex. pls use that.

File: src/services/patch_kill_list_command.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	std::string supplied_reason{};
+
+	if (params.size() > 3)
+	{
+		supplied_reason = params[2];
+	}
+
+	bool should_remove = false;
+
+	if (supplied_address[0] == '-') // e.g.  -1.2.3.4 removes it from the list, while 1.2.3.4 adds it to the list
```
Comment (momo5502) [inline src/services/patch_kill_list_command.cpp:66]: hm. a bit weird. why not add a param before that says add or remove

File: src/services/patch_kill_list_command.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	{
+		should_remove = true;
+		supplied_address = supplied_address.substr(1);
+	}
+
+
+	if (supplied_key == key)
```
Comment (momo5502) [inline src/services/patch_kill_list_command.cpp:73]: why not use a asymetric cryptography scheme for that? like ecdsa? eg. a timestamp is signed with the private key and the server verifies it? or a simple challenge

**Review by momo5502 — CHANGES_REQUESTED** [review]

File: src/services/patch_kill_list_command.cpp (hunk 2 above)
Comment (Rackover) [reply src/services/patch_kill_list_command.cpp:21]: your choice! :p 
Didnt know what you preferred as a server admin. I do prefer env on my server software.
will make it a TXT instead then

File: src/services/patch_kill_list_command.cpp (hunk 5 above)
Comment (Rackover) [reply src/services/patch_kill_list_command.cpp:66]: well idk i prefer it this way, it makes one less param. if you want an explicit fourth param placed before that says "add" or "remove" i can do it too, what do you prefer ?

Comment (momo5502) [reply src/services/patch_kill_list_command.cpp:37]: the fact that you mentioned "explicit" shows why a fourth param is better. it is explicit and therefore easier to understand

File: src/services/kill_list.cpp [hunk 7]
```diff
@@ -6,31 +6,34 @@
...
 	std::string str_address = address.to_string();
+	bool contains = false;
 
-	// Lock the list so it doesn't move while we search
-	std::lock_guard<std::recursive_mutex> _(kill_list_mutex);
-
-	if (entries.find(str_address) != entries.end())
-	{
-		auto entry = entries.at(str_address);
+	entries_container.access([str_address, &reason, &contains](const kill_list_entries& entries)
```
Comment (momo5502) [inline src/services/kill_list.cpp:11]: why are you copying the str_address? also why not just [&] ?

Comment (momo5502) [inline src/services/kill_list.cpp:11]: access takes an optional return type, so return access<bool>(...)
that way you can simply return a bool instead of exfiltrating it from the lambda

File: src/services/kill_list.cpp [hunk 8]
```diff
@@ -6,31 +6,34 @@
...
+		{
+			if (entries.find(str_address) != entries.end())
+			{
+				auto entry = entries.at(str_address);
 
-		reason = entry.reason;
-		return true;
-	}
+				reason = entry.reason;
+				contains = true;
```
Comment (momo5502) [inline src/services/kill_list.cpp:18]: return true

File: src/services/kill_list.cpp [hunk 9]
```diff
@@ -6,31 +6,34 @@
...
 
-void kill_list::add_to_kill_list(const kill_list::kill_list_entry& add)
+void kill_list::add_to_kill_list(kill_list::kill_list_entry add)
 {
-	std::lock_guard<std::recursive_mutex> _(kill_list_mutex);
+	entries_container.access([&add](kill_list_entries& entries)
+		{
+			if (entries.find(add.ip_address) == entries.end())
+			{
+				entries.emplace(add.ip_address, add);
```
Comment (momo5502) [inline src/services/kill_list.cpp:31]: std::move(add)

**Review by momo5502 — COMMENTED** [review]
will check the rest tomorrow. i'm super tired right now.

**Comment by Rackover** [comment]
I tested all the features with a custom client and they're working as intended!
- Wrong signature refuses the command
- Any other 3nd argument than "add" or "remove" refuses the command
- Kill list can be updated (elements added or removed at will)
- An existing entry can be updated (you can update the reason by adding it again with a different reason)
- Connections present on the kill list correctly get refused
- KILL.TXT is correctly updated, only when necessary

I did not do a challenge like you said, I simply used the private key to sign the message (signing the IP argument more specifically) so that only someone with the server private key can patch the list, and the worst thing an eavesdropper could do is repeating a command 🙂 There is no longer a secret key to store or anything,  Is that OK with you ?

File: src/services/kill_list.cpp [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+void kill_list::add_to_kill_list(kill_list::kill_list_entry add)
+{
+	bool any_change = false;
+	entries_container.access([&add, &any_change](kill_list_entries& entries)
+		{
+			auto existing_entry = entries.find(add.ip_address);
+			if (existing_entry == entries.end() || existing_entry->second.reason != add.reason)
+			{
+				if (existing_entry != entries.end())
```
Comment (diamante0018) [inline src/services/kill_list.cpp:32]: contains

File: src/services/kill_list.cpp [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+			while (std::getline(string_stream, line))
+			{
+				if (line[0] == '#')
+				{
+					// comments or ignored line
+					continue;
+				}
+
+				std::string ip;
+				std::string comment{};
```
Comment (diamante0018) [inline src/services/kill_list.cpp:101]: {} not needed and weird in this context

File: src/services/kill_list.cpp [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+void kill_list::write_to_disk()
+{
+	utils::io::remove_file(kill_file);
+
+	std::ostringstream stream;
+	entries_container.access([&stream, this](kill_list_entries& entries)
+	{
+		for (auto kv : entries)
```
Comment (diamante0018) [inline src/services/kill_list.cpp:144]: reference

File: src/services/kill_list.cpp [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+		{
+			auto entry = kv.second;
```
Comment (diamante0018) [inline src/services/kill_list.cpp:146]: reference

File: src/services/kill_list.cpp [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+			stream << entry.ip_address << " " << entry.reason << "\n";
+		}
+
+		utils::io::write_file(kill_file, stream.str());
+		console::info("Wrote %s to disk (%i entries)", kill_file.data(), entries.size());
```
Comment (diamante0018) [inline src/services/kill_list.cpp:151]: size in unsigned, correct format string

File: src/services/patch_kill_list_command.cpp [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+	return "patchkill";
+}
+
+// patchkill signature add/remove target_ip (ban_reason)
+void patch_kill_list_command::handle_command([[maybe_unused]] const network::address& target, const std::string_view& data)
+{
+	const utils::parameters params(data);
+	if (params.size() < 3)
+	{
+		throw execution_exception{ "Invalid parameter count" };
```
Comment (diamante0018) [inline src/services/patch_kill_list_command.cpp:22]: use () initializer

File: src/services/patch_kill_list_command.cpp [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	
+	const auto& signature = utils::cryptography::base64::decode(params[0]);
+	bool should_remove = params[1] == "remove"s;
+
+	if (!should_remove && params[1] != "add"s)
+	{
+		throw execution_exception{ "Invalid parameter #2: should be 'add' or 'remove'" };
```
Comment (diamante0018) [inline src/services/patch_kill_list_command.cpp:30]: same as above

File: src/services/patch_kill_list_command.cpp [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	auto supplied_address = params[2];
+
+	std::string supplied_reason{};
```
Comment (diamante0018) [inline src/services/patch_kill_list_command.cpp:35]: remove {}

File: src/services/patch_kill_list_command.cpp [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+	
+	const auto& signature = utils::cryptography::base64::decode(params[0]);
+	bool should_remove = params[1] == "remove"s;
+
+	if (!should_remove && params[1] != "add"s)
+	{
+		throw execution_exception{ "Invalid parameter #2: should be 'add' or 'remove'" };
+	}
+
+	auto supplied_address = params[2];
```
Comment (diamante0018) [inline src/services/patch_kill_list_command.cpp:33]: reference

File: src/services/patch_kill_list_command.cpp [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+	std::string supplied_reason{};
+
+	if (params.size() > 4)
+	{
+		for (size_t i = 3; i < params.size(); i++)
```
Comment (diamante0018) [inline src/services/patch_kill_list_command.cpp:39]: ++i

File: src/services/patch_kill_list_command.cpp [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			supplied_reason += params[i] + " ";
+		}
+	}
+
+	auto crypto_key = crypto_key::get(); 
+
+	if (!utils::cryptography::ecc::verify_message(crypto_key, supplied_address, signature))
+	{
+		throw execution_exception{ "Signature verification of the kill list patch key failed" };
```
Comment (diamante0018) [inline src/services/patch_kill_list_command.cpp:49]: same as above

File: src/std_include.hpp [hunk 21]
```diff
@@ -49,6 +49,7 @@
 #include <netdb.h>
 #include <unistd.h>
 #include <fcntl.h>
+#include <stdlib.h>
```
Comment (diamante0018) [inline src/std_include.hpp:52]: unused?

File: src/services/kill_list.cpp [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+#include "utils/io.hpp"
+#include "utils/string.hpp"
+
+bool kill_list::contains(const network::address& address, std::string& reason)
+{
+	std::string str_address = address.to_string(false);
+
+	return entries_container.access<bool>([&str_address, &reason](const kill_list_entries& entries)
+		{
+			if (entries.find(str_address) != entries.end())
```
[the remark this answers is not in the document]
Comment (Rackover) [reply src/services/kill_list.cpp:12]: does not exist in ++17

File: src/services/kill_list.cpp (hunk 10 above)
Comment (Rackover) [reply src/services/kill_list.cpp:32]: not in c++17

File: src/services/kill_list.cpp [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+	else
+	{
+		console::info("%s not kill list, not doing anything", remove.data());
+	}
+}
+
+void kill_list::reload_from_disk()
+{
+	if (utils::io::file_exists(kill_file))
+	{
```
Comment (diamante0018) [inline src/services/kill_list.cpp:83]: negate if statement. shorten the nesting. everywhere in your code possibly.

File: src/services/kill_list.hpp [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+		std::string reason;
+
+		kill_list_entry(std::string ip_address, std::string reason)
+		{
+			this->ip_address = ip_address;
+			this->reason = reason;
+		}
+	};
+
+	const std::string kill_file = "./KILL.TXT";
```
Comment (diamante0018) [inline src/services/kill_list.hpp:23]: why all caps?

**Review by diamante0018 — COMMENTED** [review]
The code generally looks okay, made some more suggestions.
Would appreciate momo5502 review as well.

File: src/services/kill_list.hpp (hunk 24 above)
Comment (Rackover) [reply src/services/kill_list.hpp:23]: does it need justification? i find it easier to read in a filesystem when important files or configuration files are all caps

Comment (Rackover) [reply src/services/kill_list.hpp:23]: plus allcap KILL.TXT is very scary, it will frighten cheaters, i guarantee

File: src/services/kill_list.cpp [hunk 25]
```diff
@@ -79,7 +79,11 @@ void kill_list::remove_from_kill_list(const std::string& remove)
 
 void kill_list::reload_from_disk()
 {
-	if (utils::io::file_exists(kill_file))
+	if (!utils::io::file_exists(kill_file))
+	{
+		console::info("Could not find %s, no kill list will be loaded.", kill_file.data());
```
Comment (diamante0018) [inline src/services/kill_list.cpp:84]: return; !!! shorten the nesting

Comment (Rackover) [reply src/services/kill_list.cpp:84]: 😠

**Comment by diamante0018** [comment]
squash this at the end, please, 19 commits :P

File: src/services/kill_list.hpp (hunk 24 above)
Comment (momo5502) [reply src/services/kill_list.hpp:23]: nah, all lowercase pls. DOS times have passed, we're in 2020

Comment (diamante0018) [reply src/services/kill_list.hpp:23]: Sweet

File: src/services/kill_list.cpp [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+	entries_container.access([&remove, &any_change](kill_list_entries& entries)
+	{
+		if (entries.erase(remove))
+		{
+			console::info("Removed %s from kill list", remove.data());
+			any_change = true;
+		}
+	});
+
+	if (any_change)
```
Comment (diamante0018) [inline src/services/kill_list.cpp:70]: could also remove else statement here by returning early (you hate now do you?)

Comment (Rackover) [reply src/services/kill_list.cpp:70]: i'm not sure this is necessary or an improvement even

File: src/services/kill_list.hpp (hunk 24 above)
Comment (Rackover) [reply src/services/kill_list.hpp]: done !

File: src/services/kill_list.cpp (hunk 26 above)
Comment (diamante0018) [reply src/services/kill_list.cpp:70]: if you purely weigh in these situations number of lines, maybe yes.

File: src/services/kill_list.cpp [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+
+void kill_list::reload_from_disk()
+{
+	if (!utils::io::file_exists(kill_file))
+	{
+		console::info("Could not find %s, no kill list will be loaded.", kill_file.data());
+		return;
+	}
+
+	std::string contents = utils::io::read_file(kill_file);
```
Comment (momo5502) [inline src/services/kill_list.cpp:88]: there is an all in one function for that -> utils::io::read_file with two args and bool return

**Review by momo5502 — CHANGES_REQUESTED** [review]

**Comment by momo5502** [comment]
pls don't mark stuff as resolved. if i commented, i want to resolve it. I'm the reviewer, not you

**Comment by Rackover** [comment]
no problem

File: src/services/patch_kill_list_command.cpp (hunk 2 above)
Comment (momo5502) [reply src/services/patch_kill_list_command.cpp]: you're creating a txt file anyways, so sticking to one scheme makes sense

**Comment by diamante0018** [comment]
I don't understand the last commit. To address 1 review, you undid everything I reviewed before: early if statement (no nesting, now there is the wrong format again for the lamba?!.
I'm like asking for the same things three times now.

**Comment by Rackover** [comment]
image
> I'll do it myself I guess since I'm probably not being listened to.
@diamante0018 

this is an extremely petty comment! i addressed it each time you mentioned it, why bother going back to edit your message 20 minutes after to add this snarky line? it's just weird

i've worked the whole day; i'm cooling off a bit this evening and having dinner. I'm sorry this 20 minutes delay is unacceptable, it's not that you're not being listened to, it's that i'm doing something else right now.

You have write access so feel free to address formatting as you did if you feel like it is an emergency. I will be back on this PR, if needed, when I have available time.  There are still some changes I need to do that I discussed with snake, so merging will not happen in the minute regardless.

**Comment by diamante0018** [comment]
I added in the last comment, not because of the 20 minutes delay, but because I started working on it myself after 20 minutes.
I don't think you had a valid reason to undo all the progress made today hence the comment.

File: src/services/kill_list.cpp [hunk 28]
```diff
@@ -12,37 +16,36 @@ bool kill_list::contains(const network::address& address, std::string& reason)
...
+	{
+		auto existing_entry = entries.find(add.ip_address_);
+		if (existing_entry == entries.end() || existing_entry->second.reason_ != add.reason_)
 		{
-			auto existing_entry = entries.find(add.ip_address);
-			if (existing_entry == entries.end() || existing_entry->second.reason != add.reason)
-			{
-				entries[add.ip_address] = std::move(add);
-				console::info("Added %s to kill list (reason: %s)", add.ip_address.data(), add.reason.data());
-				return true;
```
Comment (momo5502) [inline src/services/kill_list.cpp]: This is huge. Thanks for addressing that. I failed to spot the UAF here :O

**Comment by momo5502** [comment]
Timestamp signing is missing. If that's done it should be fine

**Comment by Rackover** [comment]
I tested it with the timestamp signing, with a duration within 5 minutes (which works) and outside that window (which gets correctly refused) 🙂  if code is good to you guys we can merge anytime

**Review by momo5502 — APPROVED** [review]
looks good to me, one @diamante0018 approved as well, it can be merged

**Review by diamante0018 — APPROVED** [review]
I filled in for louv with the last commit as she is working right now.
Used std::abs, and used params.join to avoid code duplication.
I now approve of this code.

I used <> instead of "" for the utils because it's the project-wide standard and what c++ guidelines want you to do.

**Comment by diamante0018** [comment]
Please use Github squash and merge feature. 🙏

File: src/services/kill_list.hpp [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+	{
+	public:
+		kill_list_entry() = default;
+		kill_list_entry(std::string ip_address, std::string reason);
+
+		std::string ip_address_;
+		std::string reason_;
+	};
+
+	const std::string kill_file = "./kill.txt";
```
Comment (momo5502) [inline src/services/kill_list.hpp:21]: Why is this in the header?

Comment (Rackover) [reply src/services/kill_list.hpp:21]: i thought that's where constants that were relevant to the whole class belonged, should this be somewhere else?

Comment (diamante0018) [reply src/services/kill_list.hpp:21]: top of the cpp file as constexpr const char*?

Comment (momo5502) [reply src/services/kill_list.hpp:21]: > i thought that's where constants that were relevant to the whole class belonged, should this be somewhere else?

things that are required by more than one cpp file belong in a header. it's simple as that,

Comment (Rackover) [reply src/services/kill_list.hpp:21]: alright, i didn't notice it was not used anywhere else (it might have been at some point and then not) so I should have put this in the private: section of the header. I'm also noting that on the private: section of this header, by the way, is an unused variable `secret_` that should probably be removed! 😄

**Comment by momo5502** [comment]
> Please use Github squash and merge feature. 🙏

i generally don't like squashing. you lose history. if the feature needed 50 iterations and therefore 50 commits, then that's how it is. but i want those iterations in the history. it's not like commits were expensive or anything alike.

File: src/services/kill_list.hpp (hunk 29 above)
Comment (Rackover) [reply src/services/kill_list.hpp:21]: @diamante0018 if you want to move the variables around (in CPP file as you said for instance) feel free to, i can't test compile on my machine for now so it'll have to wait for tonight otherwise 🙏

**Comment by diamante0018** [comment]
> > Please use Github squash and merge feature. 🙏
> 
> i generally don't like squashing. you lose history. if the feature needed 50 iterations and therefore 50 commits, then that's how it is. but i want those iterations in the history. it's not like commits were expensive or anything alike.

I see. I guess for these projects it's not super relevant to use squash.
I moved the thing to the cpp file.

**Comment by momo5502** [comment]
will merge it after work

**Comment by Rackover** [comment]
@snake once you're done with this PR & deployment, please send the server private key to an admin (preferrably Dss0, but Evan might need it too ultimately) so he can start the killin' 🔫     I will be explaining to him how it works & providing with a kill client so until evan is done with the bot, only dss0 (or whoever you give the pkey to) will be able to kill

**Comment by Rackover** [comment]
i will run tests with dss0 with a local master server instance on my computer to make sure everything is correct before we try any of this live, of course
